File: src/main/java/com/freya02/botcommands/internal/components/ComponentDescriptor.java
package com.freya02.botcommands.internal.components;

import com.freya02.botcommands.api.BContext;
import com.freya02.botcommands.internal.ExecutableInteractionInfo;
import com.freya02.botcommands.internal.MethodParameters;
import com.freya02.botcommands.internal.runner.MethodRunner;
import org.jetbrains.annotations.NotNull;

import java.lang.reflect.Method;

public class ComponentDescriptor implements ExecutableInteractionInfo {
	private final Method method;
	private final Object instance;
	private final MethodRunner methodRunner;
	private final MethodParameters<ComponentHandlerParameter> componentParameters;

	public ComponentDescriptor(BContext context, Object instance, Method method) {
		this.method = method;
		this.instance = instance;
		this.methodRunner = context.getMethodRunnerFactory().make(instance, method);

		this.componentParameters = MethodParameters.of(context, method, ComponentHandlerParameter::new);
	}

	@Override
	@NotNull
	public Method getMethod() {
		return method;
	}

	@Override
	@NotNull
	public MethodRunner getMethodRunner() {
		return methodRunner;
	}

	@Override
	@NotNull
	public MethodParameters<ComponentHandlerParameter> getParameters() {
		return componentParameters;
	}

	@Override
	@NotNull
	public Object getInstance() {
		return instance;
	}
}

File: src/main/kotlin/com/freya02/botcommands/internal/components/ComponentDescriptor.kt
package com.freya02.botcommands.internal.components

import com.freya02.botcommands.api.commands.builder.CustomOptionBuilder
import com.freya02.botcommands.internal.BContextImpl
import com.freya02.botcommands.internal.IExecutableInteractionInfo
import com.freya02.botcommands.internal.core.reflection.MemberEventFunction
import com.freya02.botcommands.internal.parameters.OptionParameter
import com.freya02.botcommands.internal.transformParameters
import com.freya02.botcommands.internal.utils.ReflectionUtils.nonInstanceParameters
import net.dv8tion.jda.api.events.interaction.component.GenericComponentInteractionCreateEvent
import kotlin.reflect.jvm.jvmErasure

class ComponentDescriptor(
    context: BContextImpl,
    override val function: MemberEventFunction<GenericComponentInteractionCreateEvent, *>
) : IExecutableInteractionInfo {
    override val parameters: List<ComponentHandlerParameter>

    init {
        parameters = function.nonInstanceParameters.drop(1).transformParameters(
            builderBlock = { function, parameter, declaredName ->
                val service = context.serviceContainer.peekServiceOrNull(parameter.type.jvmErasure)
                when {
                    service != null -> CustomOptionBuilder(OptionParameter.fromSelfAggregate(function, declaredName))
                    else -> ComponentHandlerOptionBuilder(OptionParameter.fromSelfAggregate(function, declaredName))
                }
            },
            aggregateBlock = { ComponentHandlerParameter(context, it) }
        )
    }
}